rbd-nbd: allow to restart daemon without destroying device

trociny opened the pull request:
If `--persist` option is specified, rbd-nbd process does not
destroy the device on exit and a new process may still reconnect
to it until `--persist-timeout` is reached.

This is supposed to be useful for live upgrades or after crash
recovering.

Signed-off-by: Mykola Golub <mgolub@suse.com>

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

lxbsz commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -983,7 +994,7 @@ static int netlink_disconnect(int index)
 
   if (!genlmsg_put(msg, NL_AUTO_PORT, NL_AUTO_SEQ, nl_id, 0, 0,
-                   NBD_CMD_DISCONNECT, 0)) {
+                   persist ? NBD_CMD_LINK_DEAD : NBD_CMD_DISCONNECT, 0)) {
```
As metioned in the mail, you shouldn't use the `NBD_CMD_LINK_DEAD` here.

lxbsz commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -1648,6 +1676,18 @@ static int parse_args(vector<const char*>& args, std::ostream *err_msg,
       cfg->set_max_part = true;
+    } else if (ceph_argparse_flag(args, i, "--persist", (char *)NULL)) {
+      cfg->persist = true;
+    } else if (ceph_argparse_witharg(args, i, &cfg->persist_timeout, err,
```
For the `persist_timeout` here you may should set up one default value if absent. 30 seocnds is recommend.

trociny replied:
It is set to 60 by default (persist_timeout is initialized with this value).

lxbsz replied:
The default IO timer in kernel for nbd is 30 seconds. So the value >= 30 should be okay.

lxbsz commented:
You'd better to get rid of relying on the pid for the `list-mapped`, and try to store the mapped info in one separate pool.

trociny commented:
Storing metadata in a pool may be a good solution in the case of more complicated daemons like nbd-runner, but I would to avoid it in the case of this "simple" daemon.

dillaman commented:
> Storing metadata in a pool may be a good solution in the case of more complicated daemons like nbd-runner, but I would to avoid it in the case of this "simple" daemon.

I'd worry about this when at cloud scale. The iSCSI gateways support an upper-bound of 4 nodes (at least right now) vs potentially thousands of nodes storing their current runtime state in the Ceph cluster.

lxbsz commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -1128,6 +1172,9 @@ static int netlink_connect(Config *cfg, struct nl_sock *sock, int nl_id, int fd,
   NLA_PUT_U64(msg, NBD_ATTR_SERVER_FLAGS, flags);
+  if (cfg->persist) {
+    NLA_PUT_U64(msg, NBD_ATTR_DEAD_CONN_TIMEOUT, cfg->persist_timeout);
+  }
```
For the `NBD_ATTR_DEAD_CONN_TIMEOUT` we'd better to set it normally for all the devices defautly. This is not only useful for the remap but also for multiple connections cases.

We can just set one default value, like 30s as default, and keep the `--persist_timeout` options to allow user to disable it or change it.

trociny replied:
rbd-nbd doesn't support multiple connections right now, so in this case it is useful for remap only.

I don't mind to always set `DEAD_CONN_TIMEOUT` but I don't like very much that `persist_timeout` option will be used for this, even if it is not `persist` mode, which may be a little confusing for users.

So I would either leave it as it is now (the timeout is set only if `--persist` is on) or change `persist_timeout` name to something more generic, still the name should suggest that it is useful for `persist` mode. I can't think anything better than just `--dead-connection-timeout`, though I am not like this very much.

dillaman replied:
Shouldn't `--persist` just be the default? We have the `rbd-nbd unmap` command which can be used to cleanly shut down a device (or cleanup a device from a crashed `rbd-nbd` process).

trociny replied:
Ack. I will try.

lxbsz replied:
That's okay for me keept the `--persist_timeout` as it is is now.

Maybe if we will set the `persist` is as default, shouldn't the `timeout` also has one default value ? Or you will also randomly could see the IO error before the new socket connection is set.

This timeout is trying to not fail the IO cmds immediately if they see all the connections are dead before their own timer expires.

trociny commented:
@dillaman I think it is in a good enough shape for preliminary review.

The commits are not squashed yet. I am going to do this later if you like my recent changes.

The new version always sets dead timeout if netlink is used, and it is always possible to reconnect if the process was killed (not
terminated with `unmap` command). And when restarting (reconnecting) the process the  `--reconnect` option for `map` command should be specified.

When reconnecting it additionally checks that the image is not currently mapped (no running process). On practice (see the test) it means that if one runs `map --reconnect` juster after `kill` command it will usually fail because the old process is not terminated yet. Do we want to improve this for this particular case? E.g. we could add a loop in rbd-nbd to retry several times if `--reconnect` is specified. Or we could add `remap` command that would kill the process (if it is still running) wait for it to terminate and then reconnect, so users would use it instead of manually killing and mapping with `--reconnect`.

Also, adding the rbd-nbd thrasher is in my TODO list. Do you have some idea how it could be implemented? I think I could add it for fsx nbd test. It could be built into fsx.c, or it could be a just workload script, similar to how we run `qemu_dynamic_features.sh` and `qemu`.

Or it could be a completely new test for rbd-nbd, e.g. mapping, mounting a fs and executing it in some way,  with the thrasher scrip running concurrently.

trociny commented:
I rebased it squashing some commits.

There is yet another question I forgot to mention earlier. In the current version I decided to use `--timeout` for both nbd request timeout and dead timeout, so we do not introduce yet another option, which I can't think of a good name for. The problem is that currently these timeouts have different default values, and if one would want to change the default for one of them via the option it would affect the other one, which probably not what one would want. I have no idea if it is bad or not.

@lxbsz Previously you were asking if `timeout` also has the default value? The default is 0, which AFAIR means no timeout (the request would hang forever), and I am not sure we want to change this to match the dead timeout (e.g. 30 or 60 sec).

dillaman commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -39,6 +40,7 @@
 #include <libnl3/netlink/genl/ctrl.h>
 #include <libnl3/netlink/genl/mngt.h>
 
+#include <experimental/filesystem>
```
Nit: we probably need to do the same awkward `__has_include` hack that lots of other files are doing in Ceph ... or create a "include/ceph_filesystem.h" helper that just does it once and pulls the filesystem namespace into "ceph::fs" (or similar)

dillaman replied:
This comment is still unaddressed

trociny replied:
@dillaman  sorry for this. It should be addressed now.

dillaman commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -104,6 +122,7 @@ static void usage()
             << "  --exclusive             Forbid writes by other clients\n"
             << "  --max_part <limit>      Override for module param max_part\n"
             << "  --nbds_max <limit>      Override for module param nbds_max\n"
+            << "  --reconnect             Reconnect to recently detached device\n"
```
Nit: just thinking, but would it perhaps make more sense to have a `map` command and a separate `attach` command to be more explicit with the option parsing? That would eventually allow support for a similar `unmap` vs `detach` command.

trociny replied:
I think I like additional `attach` + `detach` commands. Do we also want `rbd device attach|detach` wrappers? I think we could decide and add this later if it proves to be useful.

Would it make sense for `detach` to wait for the process to actually terminate? This would allow to avoid races when `detach` following by `attach` are executed.

dillaman replied:
> Do we also want `rbd device attach|detach` wrappers? I think we could decide and add this later if it proves to be useful.

I'm OK with waiting to add then to the "rbd" CLI -- it's a pretty narrow use-case.

> Would it make sense for `detach` to wait for the process to actually terminate? This would allow to avoid races when `detach` following by `attach` are executed.

If we can get "rbd-nbd" to properly wait for the `unmap` and `detach` cases, I think that would be great!

dillaman commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -300,13 +325,39 @@ class NBDServer
       dout(20) << __func__ << ": waiting for nbd request" << dendl;
 
-      int r = safe_read_exact(fd, &ctx->request, sizeof(struct nbd_request));
+      // We can not block on read because we have to check periodically
```
Nit: you can use an `EventSocket` to send a wakeup signal.

dillaman commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -749,8 +833,34 @@ class NBDListIterator {
+        continue;
+      }
+
+      int pid = atoi(entry.path().filename().c_str());
```
Nit: you could just use `boost::lexical_cast` to determine if the directory name was an integer

dillaman reviewed (commented):
> Previously you were asking if timeout also has the default value? The default is 0, which AFAIR means no timeout (the request would hang forever), and I am not sure we want to change this to match the dead timeout (e.g. 30 or 60 sec).

Maybe `--timeout` could be split out into `--io-timeout` vs `--reattach-timeout` (or similar)

lxbsz commented:
> @lxbsz Previously you were asking if `timeout` also has the default value? The default is 0, which AFAIR means no timeout (the request would hang forever), and I am not sure we want to change this to match the dead timeout (e.g. 30 or 60 sec).

No, without specifying the `--timeout`, it will be 30 seconds as default.

```
1361 static void nbd_set_cmd_timeout(struct nbd_device *nbd, u64 timeout)
1362 { 
1363         nbd->tag_set.timeout = timeout * HZ;
1364         if (timeout)          
1365                 blk_queue_rq_timeout(nbd->disk->queue, timeout * HZ);
1366         else
1367                 blk_queue_rq_timeout(nbd->disk->queue, 30 * HZ);                                                                     
1368 } 
```

Please note that the `--timeout` in rbd-nbd.cc currently is for the IO commands timeout only. The `--persist-timeout` you added before was for how long should it wait to setup the IO command timeout timer, 0 as default means no wait and fails directly.

trociny commented:
> No, without specifying the --timeout, it will be 30 seconds as default.

Indeed. I forgot about this. Then if we made the dead timeout default be 30 sec too (actually how it is in my code currently) then it does not look so bad to me to have just one option for both params.

If people still prefer separate options (`io-timeout`, `reattach-timeout`), I don't mind. But I think we couldn't just rename `timeout` to `io-timeout`. We would need to leave `timeout` as an alias and deprecate it.

dillaman commented:
> Indeed. I forgot about this. Then if we made the dead timeout default be 30 sec too (actually how it is in my code currently) then it does not look so bad to me to have just one option for both params.

I think 30 secs is fine -- but I'd vote for separate params for clarity and configurability going forward since it's hard to change once we've locked in to a combined parameter.

trociny commented:
@dillaman I think it is ready for review now.

dillaman commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -1486,6 +1683,19 @@ static int do_map(int argc, const char *argv[], Config *cfg)
+  int r = kill(cfg->pid, SIGTERM);
+  if (r == -1) {
+    r = -errno;
+    cerr << "rbd-nbd: failed to kill " << cfg->pid << ": " << cpp_strerror(r)
```
```suggestion
    cerr << "rbd-nbd: failed to terminate " << cfg->pid << ": " << cpp_strerror(r)
```

dillaman commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -1284,7 +1423,65 @@ static void run_server(Preforker& forker, NBDServer *server, bool netlink_used)
+  if (r < 0) {
+    cerr << "rbd-nbd: failed to poll rbd-nbd process: " << cpp_strerror(r)
+         << std::endl;
+    goto done;
```
Nit: leaking fd in this case

trociny replied:
Ah, actually the `done` label was introduced to avoid the leak, but I jut put it in a wrong place. Thanks.

dillaman commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -1284,7 +1423,65 @@ static void run_server(Preforker& forker, NBDServer *server, bool netlink_used)
 
-static int do_map(int argc, const char *argv[], Config *cfg)
+#ifndef __NR_pidfd_open
+#define __NR_pidfd_open 434   /* System call # on most architectures */
```
Nit: if there is no glibc wrapper for it and it's not defined, I think you should just fallback to the `/proc/` approach ... or just always open the `/proc/<pid>` path to avoid the fallback path.

dillaman commented on src/tools/rbd_nbd/rbd-nbd.cc:
```diff
@@ -300,13 +335,36 @@ class NBDServer
-      int r = safe_read_exact(fd, &ctx->request, sizeof(struct nbd_request));
+      int r = poll(poll_fds, 2, -1);
+      if (r < 0) {
+        derr << "failed to poll nbd: " << cpp_strerror(r) << dendl;
```
Nit: ignore `errno == EINTR` errors and make sure you grab the error from `errno` instead of attempting to print `r`

trociny commented:
updated

dillaman reviewed (approved):
lgtm

trociny commented:
@dillaman pushed rbd-nbd man update.

dillaman commented:
jenkins test make check

dillaman commented:
http://qa-proxy.ceph.com/teuthology/jdillaman-2020-10-08_15:40:00-rbd-wip-jd-testing-distro-basic-smithi/5507063/teuthology.log

> 2020-10-08T19:58:36.848 INFO:tasks.workunit.client.0.smithi179.stderr:++ sudo -nE /usr/bin/rbd-nbd map rbd/testrbdnbd27875
> 2020-10-08T19:58:37.033 INFO:tasks.workunit.client.0.smithi179.stderr:+ DEV=/dev/nbd0
> 2020-10-08T19:58:37.033 INFO:tasks.workunit.client.0.smithi179.stderr:+ get_pid
> 2020-10-08T19:58:37.033 INFO:tasks.workunit.client.0.smithi179.stderr:+ local ns=
> 2020-10-08T19:58:37.034 INFO:tasks.workunit.client.0.smithi179.stderr:++ rbd-nbd --format xml list-mapped
> 2020-10-08T19:58:37.034 INFO:tasks.workunit.client.0.smithi179.stderr:++ xmlstarlet sel -t -v '//devices/device[pool='\''rbd'\''][namespace='\'''\''][image='\''testrbdnbd27875'\''][device='\''/dev/nbd0'\'']/id'
> 2020-10-08T19:58:37.051 INFO:tasks.workunit.client.0.smithi179.stderr:+ PID=
> 2020-10-08T19:58:37.051 INFO:tasks.workunit.client.0.smithi179.stderr:+ cleanup

trociny commented:
> http://qa-proxy.ceph.com/teuthology/jdillaman-2020-10-08_15:40:00-rbd-wip-jd-testing-distro-basicsmithi/5507063/teuthology.log

Hm, in the rbd-nbd log I see:
```
2020-10-08T19:58:36.885+0000 7f6e066a3e80  0 ceph version 16.0.0-6217-g6c7aaa5a (6c7aaa5a9ddbf8da8ca97654fe08060034ee5838) pacific (dev), process rbd-nbd, pid 28799
2020-10-08T19:58:36.886+0000 7f6e066a3e80  0 pidfile_write: ignore empty --pid-file
2020-10-08T19:58:37.027+0000 7f6dd37fe700  0 rbd-nbd: reader_entry: terminate received
```
"reader_entry: terminate received" means the `poll` returned there was something to read from `terminate_event_fd`, which is supposed to be only if the signal handler has sent the event socket notify. But if it had been really the case and the `handle_signal` had been called I would have expected  "*** Got signal " message in the log, which I don't see...

trociny commented:
Adding some diagnostic and experimenting on teuthology I tracked this further.

So, in the problem case, `ioctl(nbd, NBD_DO_IT)` returns immediately with `EBUSY` error and `rbd-nbd` terminates. And I was able to reproduce it on a teuthology node just running `rbd-nbd map; rbd-nbd unmap` in a loop. In `dmesg` one can see then:
```
[ 3244.161347] block nbd0: shutting down sockets
[ 3244.166032] blk_update_request: I/O error, dev nbd0, sector 130488 op 0x0:(READ) flags 0x0 phys_seg 5 prio class 0
[ 3244.176741] blk_update_request: I/O error, dev nbd0, sector 130488 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 0
[ 3244.187441] Buffer I/O error on dev nbd0, logical block 130488, async page read
[ 3244.195087] blk_update_request: I/O error, dev nbd0, sector 130489 op 0x0:(READ) flags 0x0 phys_seg 7 prio class 0
[ 3244.205786] Buffer I/O error on dev nbd0, logical block 130489, async page read
[ 3244.213461] Buffer I/O error on dev nbd0, logical block 130490, async page read
[ 3244.221117] Buffer I/O error on dev nbd0, logical block 130491, async page read
[ 3244.228769] Buffer I/O error on dev nbd0, logical block 130492, async page read
[ 3244.236416] Buffer I/O error on dev nbd0, logical block 130493, async page read
[ 3244.244055] Buffer I/O error on dev nbd0, logical block 130494, async page read
[ 3244.251685] Buffer I/O error on dev nbd0, logical block 130495, async page read
[ 3244.259343] blk_update_request: I/O error, dev nbd0, sector 0 op 0x0:(READ) flags 0x0 phys_seg 8 prio class 0
[ 3244.269599] Buffer I/O error on dev nbd0, logical block 0, async page read
[ 3244.276818] Buffer I/O error on dev nbd0, logical block 1, async page read
[ 3244.284038] blk_update_request: I/O error, dev nbd0, sector 4096 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 0
[ 3244.294569] blk_update_request: I/O error, dev nbd0, sector 0 op 0x0:(READ) flags 0x0 phys_seg 8 prio class 0
[ 3244.304846] blk_update_request: I/O error, dev nbd0, sector 4096 op 0x0:(READ) flags 0x0 phys_seg 8 prio class 0
[ 3244.315384] blk_update_request: I/O error, dev nbd0, sector 0 op 0x0:(READ) flags 0x0 phys_seg 8 prio class 0
[ 3244.325715] blk_update_request: I/O error, dev nbd0, sector 0 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 0
[ 3244.325733] blk_update_request: I/O error, dev nbd0, sector 1 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 0
[ 3244.325860] Dev nbd0: unable to read RDB block 0
[ 3244.325944]  nbd0: unable to read partition table
```
I believe this is some kind of a race possible when `rbd-nbd map` is run just after `rbd-nbd unmap`, and we started observe it with my PR because I got rid of `unmap_device` function which always was sleeping at least 0.5 sec before returning. When I add `sleep 0.5` into my map-unmap loop I can't reproduce the issue.

dillaman commented:
> So, in the problem case, `ioctl(nbd, NBD_DO_IT)` returns immediately with `EBUSY` error and `rbd-nbd` terminates. And I was able to reproduce it on a teuthology node just running `rbd-nbd map; rbd-nbd unmap` in a loop. In `dmesg` one can see then:

So this was using the `ioctl` interface where it needs to test each `/dev/nbdX` device? I wonder if we can just put better logic in place to pick a new device if you hit an `EBUSY` when mapping via `ioctl`. Those `dmesg` IO errors definitely seem like `udev`-related partition probing, though.

trociny commented:
> So this was using the `ioctl` interface where it needs to test each `/dev/nbdX` device?

Yes.

> I wonder if we can just put better logic in place to pick a new device if you hit an `EBUSY` when mapping via ioctl.

The problem is that it returns `EBUSY` only at `ioctl(nbd, NBD_DO_IT)`, that is after we successfully opened the device and returned its name to the user and going to block waiting for disconnect. Retry with another device from this stage looks tricky.

dillaman commented:
I'll have to look deeper at the RHEL 8.1 code since I don't immediately see how `EBUSY` is ever returned on the `ioctl` path in the upstream kernel.

trociny commented:
Thanks, I will play with it.

lxbsz commented:
> I'll have to look deeper at the RHEL 8.1 code since I don't immediately see how `EBUSY` is ever returned on the `ioctl` path in the upstream kernel. I just wonder if it's `udev` locking the device while it's playing around that is the true source of the `EBUSY` and therefore you could probably run a `udevadm settle` (or similar) and try to `NBD_DO_IT` again). If you can't repeat the failure when `udevd` is not running I think it's a good start.

In upstream the `EBUSY` was returned by `__nbd_ioctl()` --> `nbd_start_device_ioctl()` --> `nbd_start_device()`:
```

1250         if (nbd->task_recv)   
1251                 return -EBUSY;
```

That means the `nbd` device has already been started and used by another task.

lxbsz commented:
> Adding some diagnostic and experimenting on teuthology I tracked this further.
> 
> So, in the problem case, `ioctl(nbd, NBD_DO_IT)` returns immediately with `EBUSY` error and `rbd-nbd` terminates. And I was able to reproduce it on a teuthology node just running `rbd-nbd map; rbd-nbd unmap` in a loop. In `dmesg` one can see then:
> 
> ```
> [ 3244.161347] block nbd0: shutting down sockets
> [ 3244.166032] blk_update_request: I/O error, dev nbd0, sector 130488 op 0x0:(READ) flags 0x0 phys_seg 5 prio class 0
> [ 3244.176741] blk_update_request: I/O error, dev nbd0, sector 130488 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 0
> [ 3244.187441] Buffer I/O error on dev nbd0, logical block 130488, async page read
> [ 3244.195087] blk_update_request: I/O error, dev nbd0, sector 130489 op 0x0:(READ) flags 0x0 phys_seg 7 prio class 0
> [ 3244.205786] Buffer I/O error on dev nbd0, logical block 130489, async page read
> [ 3244.213461] Buffer I/O error on dev nbd0, logical block 130490, async page read
> [ 3244.221117] Buffer I/O error on dev nbd0, logical block 130491, async page read
> [ 3244.228769] Buffer I/O error on dev nbd0, logical block 130492, async page read
> [ 3244.236416] Buffer I/O error on dev nbd0, logical block 130493, async page read
> [ 3244.244055] Buffer I/O error on dev nbd0, logical block 130494, async page read
> [ 3244.251685] Buffer I/O error on dev nbd0, logical block 130495, async page read
> [ 3244.259343] blk_update_request: I/O error, dev nbd0, sector 0 op 0x0:(READ) flags 0x0 phys_seg 8 prio class 0
> [ 3244.269599] Buffer I/O error on dev nbd0, logical block 0, async page read
> [ 3244.276818] Buffer I/O error on dev nbd0, logical block 1, async page read
> [ 3244.284038] blk_update_request: I/O error, dev nbd0, sector 4096 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 0
> [ 3244.294569] blk_update_request: I/O error, dev nbd0, sector 0 op 0x0:(READ) flags 0x0 phys_seg 8 prio class 0
> [ 3244.304846] blk_update_request: I/O error, dev nbd0, sector 4096 op 0x0:(READ) flags 0x0 phys_seg 8 prio class 0
> [ 3244.315384] blk_update_request: I/O error, dev nbd0, sector 0 op 0x0:(READ) flags 0x0 phys_seg 8 prio class 0
> [ 3244.325715] blk_update_request: I/O error, dev nbd0, sector 0 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 0
> [ 3244.325733] blk_update_request: I/O error, dev nbd0, sector 1 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 0
> [ 3244.325860] Dev nbd0: unable to read RDB block 0
> [ 3244.325944]  nbd0: unable to read partition table
> ```
> 
> I believe this is some kind of a race possible when `rbd-nbd map` is run just after `rbd-nbd unmap`, and we started observe it with my PR because I got rid of `unmap_device` function which always was sleeping at least 0.5 sec before returning. When I add `sleep 0.5` into my map-unmap loop I can't reproduce the issue.

Checked the unmap code in kernel, the nbd device `put` maybe deferred, it will just decrease the reference count and returned, but won't make sure that the nbd device is acutally back into a free state if there has some other tasks are using the nbd device. 

So next time when you try to connect the same nbd device it will return EBUSY.

lxbsz commented:
```
1470 static int nbd_open(struct block_device *bdev, fmode_t mode)
1471 { 
1472         struct nbd_device *nbd;
1473         int ret = 0;
1474   
1475         mutex_lock(&nbd_index_mutex);       
1476         nbd = bdev->bd_disk->private_data;  
1477         if (!nbd) {           
1478                 ret = -ENXIO;
1479                 goto out;     
1480         }
1481         if (!refcount_inc_not_zero(&nbd->refs)) {
1482                 ret = -ENXIO; 
1483                 goto out;     
1484         }
```

Maybe you can just try to open the nbd device if the /dev/nbdX device is already existed, and then it will return a `-ENXIO` errno if the nbd device is still used by other tasks.

trociny commented:
> Maybe you can just try to open the nbd device if the /dev/nbdX device is already existed, and then it will return a -ENXIO errno if the nbd device is still used by other tasks.

I am not quite sure I understand what you mean here. The `rbd-nbd` (`try_ioctl_setup`) already does what you seem to suggest: it tries to open a device and proceeds only if the open succeeded.

lxbsz commented:
> > Maybe you can just try to open the nbd device if the /dev/nbdX device is already existed, and then it will return a -ENXIO errno if the nbd device is still used by other tasks.
> 
> I am not quite sure I understand what you mean here. The `rbd-nbd` (`try_ioctl_setup`) already does what you seem to suggest: it tries to open a device and proceeds only if the open succeeded.

Forget it, this seems won't work.

lxbsz commented:
Checked the nbd driver code again, there is only one case will return the `EBUSY` errno if it's really failed from `NBD_DO_IT` ioctl, that is the previous `unmap` is not finished and the last `nbd_config_put()` is not called yet, the `nbd_config_put()` will clear the `task_recv`.

There has 2 places will call the `nbd_config_put()` for the ioctl method, in `nbd_release()` and after each socket is closed. The socket closing ones maybe deferred.

Have you tried add one `sleep` in your `rbd-nbd map; rbd-nbd unmap` loop before the `map` ? With somthing like this if it really work well, so this should be root cause.

trociny commented:
> Have you tried add one sleep in your rbd-nbd map; rbd-nbd unmap loop before the map ? With somthing like this if it really works well, so this should be root cause.

Yes I tried to add the sleep and it helped (see my initial comment [1]). The question is how to make `rbd-nbd map` work reliable without the sleep. Note, it is not only the test problem, you may face this problem in reality, when `unmap` happens immediately after `map`, not necessary for the same image.

[1] https://github.com/ceph/ceph/pull/37273#issuecomment-707641917

lxbsz commented:
> > Have you tried add one sleep in your rbd-nbd map; rbd-nbd unmap loop before the map ? With somthing like this if it really works well, so this should be root cause.
> 
> Yes I tried to add the sleep and it helped (see my initial comment [1]). The question is how to make `rbd-nbd map` work reliable without the sleep. Note, it is not only the test problem, you may face this problem in reality, when `unmap` happens immediately after `map`, not necessary for the same image.
> 
> [1] [#37273 (comment)](https://github.com/ceph/ceph/pull/37273#issuecomment-706672445)

We can use the `nbd_dbg_tasks` sysfs file to get the task `pid`:

```
static int nbd_dbg_tasks_show(struct seq_file *s, void *unused)
{
        struct nbd_device *nbd = s->private;

        if (nbd->task_recv)
                seq_printf(s, "recv: %d\n", task_pid_nr(nbd->task_recv));                                                                 

        return 0;
}
```
And then check if the correspending task, if it does not exist, then we can assume that this task should be a stale one, then the `map` could wait for a while and retry. Or if the task existes, the `map` should fail. Does this make sense ?

trociny commented:
Interestingly, I can't reproduce it with `--try-netlink`...

dillaman commented:
It looks like the `unmap` race is back [1]:

```
2020-10-14T00:51:39.979 INFO:tasks.workunit.client.0.smithi055.stderr:+ sudo -nE /usr/bin/rbd-nbd unmap rbd/ns/testrbdnbd27664@snap
2020-10-14T00:51:40.035 INFO:tasks.workunit.client.0.smithi055.stderr:+ rbd-nbd list-mapped
2020-10-14T00:51:40.085 INFO:tasks.workunit.client.0.smithi055.stderr:+ expect_false grep '^28584\b'
2020-10-14T00:51:40.085 INFO:tasks.workunit.client.0.smithi055.stderr:+ grep '^28584\b'
2020-10-14T00:51:40.085 INFO:tasks.workunit.client.0.smithi055.stderr:+ return 1
2020-10-14T00:51:40.086 INFO:tasks.workunit.client.0.smithi055.stderr:+ return 1
2020-10-14T00:51:40.086 INFO:tasks.workunit.client.0.smithi055.stderr:+ cleanup
```

[1] http://qa-proxy.ceph.com/teuthology/jdillaman-2020-10-13_18:49:11-rbd-wip-jd-testing-distro-basic-smithi/5524291/teuthology.log

lxbsz commented:
> > And then check if the correspending task, if it does not exist, then we can assume that this task should be a stale one, then the map could wait for a while and retry. Or if the task existes, the map should fail. Does this make sense ?
> 
> Do you mean trying to access "/sys/kernel/debug/nbd/nbdX/tasks" file, and fail (retry with another device) if it succeeded?
> 

```
# cat /sys/kernel/debug/nbd/nbd0/tasks
recv: 502308

# ps -aux |grep nbd
root      502308  0.1  0.2 1682264 34532 pts/0   Sl   09:21   0:00 ./bin/rbd-nbd map cephfs.a.data/block0

```

When do mapping( for example the /dev/nbdX) and if it is failed, then check the `/sys/kernel/debug/nbd/nbdX/tasks` file, if the PID info is exist and the corresponding process does not exist, then the map should wait and retry.

Or if the corresponding process is alive, the map should fail directly with the errors that the nbdX device already mapped or used by another task.

For the ioctl method, the `map` process will stuck in `ioctl(nbd, NBD_DO_IT)` and will wait until another `unmap` process sends the disconnect command to ternimate it:

The `map` process will stuck and wait here in kernel space:
```
static int nbd_start_device_ioctl(struct nbd_device *nbd, struct block_device *bdev)
{
        ...
        ret = wait_event_interruptible(config->recv_wq,                                                                                   
                                         atomic_read(&config->recv_threads) == 0);
        if (ret)
                sock_shutdown(nbd);
        ...
}

```
When the `unmap` has sent the DISCONNECT comand, and when the last socket thread is exiting, it will do:
```
static void recv_work(struct work_struct *work)
{
        ...
        atomic_dec(&config->recv_threads);
        wake_up(&config->recv_wq);
        nbd_config_put(nbd);      
        ...
}
```
It will:
1), decrease the `config->recv_threads` reference
2), then wakeup the `map` process to terminate
3), scheduled out
4), at last it will do the `nbd_config_put`, which will decrese the `config_refs`.

There has a gap between step2 and step4, during which the above thread maybe scheduled out. But once the step2 done, the `map` process could already finished and exited, though it called the `close(nbd)`, which will also call the  `nbd_config_put` but without the above step4, it won't clearing the `nbd->task_recv`.

That means the `map` process could exit without clearing the `nbd->task_recv` in time.



> It looks a little hackish but I would probably be ok with this, if `EBUSY` was really returned due to not still released device, which I am not sure right now. Because I just tried @dillaman's suggestion to repeat the experiment when `udevd` is stopped, and I failed to reproduce the failure, even though the map/unmap loop was 10 times longer. I think it is still does not 100% guarantee it was due to `udevd` but it looks very likely to me.

trociny commented:
@lxbsz The idea to use `/sys/kernel/debug/nbd/nbdX/tasks` looks interesting. I will probably evaluate it in the future. Though I a bit worry to use `/sys/kernel/debug/` fs, as I am not sure how widely it is enabled. And interestingly, I just tried to reattach the rbd-nbd process and see that the pid did not changed in `tasks`, i.e. the same problem as with `/sys/block/nbd0/pid`.

Taking that this unmap/map race is actually out of the scope of this PR, I just added it to my TODO and modified `qa/workunits/rbd/rbd-nbd.sh`, returning back `unmap_device` and modifying it a little. The teuthology run [1].

[1] https://pulpito.ceph.com/trociny-2020-10-14_08:08:47-rbd-wip-mgolub-testing-1-distro-basic-smithi/

lxbsz commented:
Yeah, agree, let's raise a new PR to fix this issue.

lxbsz commented:
Post one patch in the NBD driver to fix it: https://patchwork.kernel.org/project/linux-block/patch/20201014024514.112822-1-xiubli@redhat.com

lxbsz commented:
> Post one patch in the NBD driver to fix it: https://patchwork.kernel.org/project/linux-block/patch/20201014024514.112822-1-xiubli@redhat.com

It's already applied to the upstream.

dillaman commented:
> > [](https://patchwork.kernel.org/project/linux-block/patch/20201014024514.112822-1-xiubli@redhat.com)
> 
> It's already applied to the upstream. But for the old kernels which may won't include this fixing we still need one workaroud to fix it in rbd-nbd.

Looking at that patch, wouldn't there be a race now? The `nbd_config_put` is applied and then you deference the `config` structure that you just decremented the reference against so in theory it could be freed, right?

lxbsz commented:
> > > 
> > 
> > 
> > It's already applied to the upstream. But for the old kernels which may won't include this fixing we still need one workaroud to fix it in rbd-nbd.
> 
> Looking at that patch, wouldn't there be a race now?

Yeah, with that patch the race should be gone. I didn't find any other race could cause this issue in kernel space.

> The `nbd_config_put` is applied and then you deference the `config` structure that you just decremented the reference against so in theory it could be freed, right?

Yeah, IMO the old code means to do:

1, disconenct all the sockets
2, notify `ioctl(nbd, NBD_DO_IT)` to continue, which will is the `map` process.
3, call `nbd_config_put` to decrease the config's reference when disconnecting the sockets.
4, when the `map` process is exiting, it will call close("/dev/nbdX"), which will call the  `nbd_config_put` too.

The Step1 is the `unmap` process.
The Step2 and Step3 are in a separate workqueue thread which will be waked up by `unmap`.
The Step4 is the `map` process.

We expect just in step4 when the `map` process is exiting, the `config` should be freed.

But the code sometime won't work as expected as above. The step3 in workqueue maybe deferred and be called after step4. Then the `config` will be freed in step3. That means the `map` process exits but leaving the `config` not be freed. This will lead the nbd->config still holding the stale process's info.

In this case both the `unmap` and `map` processes exit, if before the workqueue thread finishes it call a new `map` we will hit this issue.

dillaman commented:
> In this case both the `unmap` and `map` processes exit, if before the workqueue thread finishes we call a new `map` we will hit this issue.

Right, but with your `nbd` driver change, is it now possible for the `nbd->config` struct to be freed after the call to `nbd_config_put(nbd)` and therefore the dereference of `config->recv_wq` would be invalid a couple lines later (or is there something else still holding a reference to `nbd->config` so it's guaranteed to be still valid - but potentially leading to the same race condition issue)?

lxbsz commented:
> > In this case both the `unmap` and `map` processes exit, if before the workqueue thread finishes we call a new `map` we will hit this issue.
> 
> Right, but with your `nbd` driver change, is it now possible for the `nbd->config` struct to be freed after the call to `nbd_config_put(nbd)` and therefore the dereference of `config->recv_wq` would be invalid a couple lines later (or is there something else still holding a reference to `nbd->config` so it's guaranteed to be still valid - but potentially leading to the same race condition issue)?

For `ioctl` method, the `/dev/nbdX` device must be opened first then it can do the `ioctl()`. And the `open()` will hold it.

lxbsz commented:
> > In this case both the `unmap` and `map` processes exit, if before the workqueue thread finishes we call a new `map` we will hit this issue.
> 
> Right, but with your `nbd` driver change, is it now possible for the `nbd->config` struct to be freed after the call to `nbd_config_put(nbd)` and therefore the dereference of `config->recv_wq` would be invalid a couple lines later (or is there something else still holding a reference to `nbd->config` so it's guaranteed to be still valid - but potentially leading to the same race condition issue)?

More detail:

The kernel space fixing:
```
+	nbd_config_put(nbd);
 	atomic_dec(&config->recv_threads);
 	wake_up(&config->recv_wq);
-	nbd_config_put(nbd);
```
After the line `+ nbd_config_put(nbd);`, will only the opened file descripter hold the `nbd->config` reference. And only after the `wake_up()` above will the `map` process could continue and after which it will `close()` the opened file descripter to call the last `nbd_config_put(nbd)`. 

So the race you mentioned shouldn't happen.

dillaman commented:
> So the race you mentioned shouldn't happen.

... but what if the `recv_work` function breaks out of the `while(1)` loop while there isn't an active `ioctl` (i.e. `rbd-nbd` crashes)?

lxbsz commented:
> > So the race you mentioned shouldn't happen.
> 
> ... but what if the `recv_work` function breaks out of the `while(1)` loop while there isn't an active `ioctl` (i.e. `rbd-nbd` crashes)?

IMO, it shouldn't the `unmap` will make the `recv_work` to break the `while()` loop.

This case should happen when:

If the `map` daemon crashes, it will close all the opened fds, including the socket peers. If the `/dev/nbdX` device opened fd is closed just before the socket fds. Right ?

lxbsz commented:
> > > So the race you mentioned shouldn't happen.
> > 
> > 
> > ... but what if the `recv_work` function breaks out of the `while(1)` loop while there isn't an active `ioctl` (i.e. `rbd-nbd` crashes)?
> 
> This case should happen when:
> 
> If the `map` daemon crashes, it will close all the opened fds, including the socket peers. If the `/dev/nbdX` device opened fd is closed just before the socket fds. Right ?

```
git diff drivers/block/nbd.c
diff --git a/drivers/block/nbd.c b/drivers/block/nbd.c
index edf8b632e3d2..0fd9e635e5ae 100644
--- a/drivers/block/nbd.c
+++ b/drivers/block/nbd.c
@@ -1320,6 +1320,8 @@ static int nbd_start_device_ioctl(struct nbd_device *nbd, struct block_device *b
        if (ret)
                return ret;
 
+       refcount_inc(&nbd->config_refs);
+
        if (max_part)
                bdev->bd_invalidated = 1;
        mutex_unlock(&nbd->config_lock);
@@ -1336,6 +1338,8 @@ static int nbd_start_device_ioctl(struct nbd_device *nbd, struct block_device *b
                ret = 0;
        if (test_bit(NBD_RT_TIMEDOUT, &config->runtime_flags))
                ret = -ETIMEDOUT;
+
+       nbd_config_put(nbd);
        return ret;
 }
 
```

This should fix this case.

lxbsz commented:
> > > So the race you mentioned shouldn't happen.
> > 
> > 
> > ... but what if the `recv_work` function breaks out of the `while(1)` loop while there isn't an active `ioctl` (i.e. `rbd-nbd` crashes)?
> 
> This case should happen when:
> 
> If the `map` daemon crashes, it will close all the opened fds, including the socket peers. If the `/dev/nbdX` device opened fd is closed just before the socket fds. Right ?

Checked it again carefully, it seems it won't happen.

For example:

In case if the `rbd-nbd map ...` daemon crashes before the `ioctl(nbd, NBD_DO_IT)`, the `recv_work` threads won't be fired, so it won't happen.

In case if the `rbd-nbd map ...` daemon crashes after the `ioctl(nbd, NBD_DO_IT)`, the `recv_work` threads should have finished, so it won't happen too.   

In case if the `rbd-nbd map ...` daemon crashes during the `ioctl(nbd, NBD_DO_IT)` waiting in 

```
1327         mutex_unlock(&nbd->config_lock);    
1328         ret = wait_event_interruptible(config->recv_wq,
1329                                          atomic_read(&config->recv_threads) == 0);
1330         if (ret)
1331                 sock_shutdown(nbd);
1332         flush_workqueue(nbd->recv_workq);
```
I have no idea in what case it will.

There has another case that the above singal interruptible wait could be waked up by a singal, but it should be okay, becuase the Line#1331 and Line#1332 will make sure all the `recv_works` threads are completed.

Is there any other case that I missed ?

trociny commented:
> In case if the `rbd-nbd map ...` daemon crashes during the `ioctl(nbd, NBD_DO_IT)` waiting in
> I have no idea in what case it will. If it really will, the kernel should crash any way, so we can ignore it.

I am not sure I quite understand. May be I just missed the point, but to me it is actually the most common case when `rbd-nbd` will crash: the main thread is waiting on `ioctl(nbd, NBD_DO_IT)` and some of the worker threads crashes, e.g. on a librbd assertion failure.

Anyway, if even the patch is really expected to be safe for all cases (and note, it is not only rbd-nbd), I would expect a comment in the source code explaining why it is safe to dereference `config` after it is released in this particular case, because for a casual reader it looks unsafe. And even if it is safe now it might become unsafe after some future change in the code. And actually I would try to avoid such things at all.

lxbsz commented:
> > In case if the `rbd-nbd map ...` daemon crashes during the `ioctl(nbd, NBD_DO_IT)` waiting in
> > I have no idea in what case it will. If it really will, the kernel should crash any way, so we can ignore it.
> 
> I am not sure I quite understand. May be I just missed the point, but to me it is actually the most common case when `rbd-nbd` will crash: the main thread is waiting on `ioctl(nbd, NBD_DO_IT)` and some of the worker threads crashes, e.g. on a librbd assertion failure.
> 
> Anyway, if even the patch is really expected to be safe for all cases (and note, it is not only rbd-nbd), I would expect a comment in the source code explaining why it is safe to dereference `config` after it is released in this particular case, because for a casual reader it looks unsafe. And even if it is safe now it might become unsafe after some future change in the code. And actually I would try to avoid such things at all.

Have tested this case, I didn't see any issue yet. Except the new ones I hit and have post 2 patches to the kernel mail list and have cc you.

lxbsz commented:
> > > In case if the `rbd-nbd map ...` daemon crashes during the `ioctl(nbd, NBD_DO_IT)` waiting in
> > > I have no idea in what case it will. If it really will, the kernel should crash any way, so we can ignore it.
> > 
> > 
> > I am not sure I quite understand. May be I just missed the point, but to me it is actually the most common case when `rbd-nbd` will crash: the main thread is waiting on `ioctl(nbd, NBD_DO_IT)` and some of the worker threads crashes, e.g. on a librbd assertion failure.
> > Anyway, if even the patch is really expected to be safe for all cases (and note, it is not only rbd-nbd), I would expect a comment in the source code explaining why it is safe to dereference `config` after it is released in this particular case, because for a casual reader it looks unsafe. And even if it is safe now it might become unsafe after some future change in the code. And actually I would try to avoid such things at all.
> 
> Have tested this case, I didn't see any issue yet. Except the new ones I hit and have post 2 patches to the kernel mail list and have cc you.

> > > In case if the `rbd-nbd map ...` daemon crashes during the `ioctl(nbd, NBD_DO_IT)` waiting in
> > > I have no idea in what case it will. If it really will, the kernel should crash any way, so we can ignore it.
> > 
> > 
> > I am not sure I quite understand. May be I just missed the point, but to me it is actually the most common case when `rbd-nbd` will crash: the main thread is waiting on `ioctl(nbd, NBD_DO_IT)` and some of the worker threads crashes, e.g. on a librbd assertion failure.
> > Anyway, if even the patch is really expected to be safe for all cases (and note, it is not only rbd-nbd), I would expect a comment in the source code explaining why it is safe to dereference `config` after it is released in this particular case, because for a casual reader it looks unsafe. And even if it is safe now it might become unsafe after some future change in the code. And actually I would try to avoid such things at all.
> 
> Have tested this case, I didn't see any issue yet. Except the new ones I hit and have post 2 patches to the kernel mail list and have cc you.

Checked the vfs related code today, the vfs will guarantee that this won't happen, otherwise there can fundamental
issue wrt race between open/close/ioctl. Before calling ioctl(), fget() is called on this FD, and ->release() is only called when the fd refcount drops to zero.

More detail, please see `SYSCALL_DEFINE3(ioctl, ...)` in `fs/ioctl.c`. Or there will be a double lock issue for the nbd ioctl too, please see the patch https://www.spinics.net/lists/linux-block/msg60397.html, which makes no sense.

lxbsz commented:
> > > In case if the `rbd-nbd map ...` daemon crashes during the `ioctl(nbd, NBD_DO_IT)` waiting in
> > > I have no idea in what case it will. If it really will, the kernel should crash any way, so we can ignore it.
> > 
> > 
> > I am not sure I quite understand. May be I just missed the point, but to me it is actually the most common case when `rbd-nbd` will crash: the main thread is waiting on `ioctl(nbd, NBD_DO_IT)` and some of the worker threads crashes, e.g. on a librbd assertion failure.
> > Anyway, if even the patch is really expected to be safe for all cases (and note, it is not only rbd-nbd), I would expect a comment in the source code explaining why it is safe to dereference `config` after it is released in this particular case, because for a casual reader it looks unsafe. And even if it is safe now it might become unsafe after some future change in the code. And actually I would try to avoid such things at all.
> 
> Have tested this case, I didn't see any issue yet. Except the new ones I hit and have post 2 patches to the kernel mail list and have cc you.

Checked the vfs related code today, before calling ioctl(), fget() is called on this FD, and ->release() is only called when the fd refcount drops to zero. More detail, please see SYSCALL_DEFINE3(ioctl, ...) in fs/ioctl.c. Or there will be a double lock issue for the nbd ioctl too, please see the patch https://www.spinics.net/lists/linux-block/msg60397.html, which makes no sense.

For this issue that we are worrying about here could be guarrenteed by flushing the workqueue before the `nbd_ioctl()` exising, to make sure the recv_work() thread has been finished when the nbd_release() is called.
